### ISSUE TITLE
SVHN dataset

### ISSUE THREAD

**Issue by bamos** [pytorch/vision#59]
Hi, I'm interested in adding format 2 (the classification task) of the SVHN dataset: http://ufldl.stanford.edu/housenumbers/

I'm planning on looking at the current dataset code in this repo and will add a similar `svhn.py` with a stubbed option for format 1. Let me know if there are any fundamental issues I may hit when trying to add this, or also if anybody has an uncommitted version if this that they can add.

-Brandon.

**Comment by soumith** [comment]
no uncommitted SVHN version afaik.

Right now, the dataset is distributed as `.mat` files, which I dont know if there's an obvious way to read in Python.
If you want to repackage the dataset as a pickle or hdf5 and upload it to pytorch's s3 bucket, let me know.

p.s.: found scipy.io that might solve the loading: https://docs.scipy.org/doc/scipy-0.18.1/reference/tutorial/io.html

**Comment by tomsercu** [comment]
Hi @bamos, Did you get around to doing this? I'll look at this tomorrow morning but if you did this already, would you mind sharing a gist? Thanks! Tom

**Comment by soumith** [comment]
Just checking in on this, is anyone working on this, or can I assign it to @uridah ?

**Comment by soumith** [comment]
Some pointers for Uridah:

SVHN Dataset: http://ufldl.stanford.edu/housenumbers/
The classification dataset has pre-cropped numbers, ranging from 0 to 9.
The task is to write a data loader similar to [CIFAR-10](https://github.com/pytorch/vision/blob/master/torchvision/datasets/cifar.py#L15-L155) that can load the SVHN dataset.
SVHN has the train, val, extra parts in the dataset, and you would want to have the user select which subset they want via a keyword argument in the constructor.
It is shipped as Matlab .mat files, that can be read by Scipy: https://docs.scipy.org/doc/scipy-0.18.1/reference/tutorial/io.html

**Comment by tomsercu** [comment]
I didn't get to this, so go ahead. 
`scipy.io.loadmat` works fine, and one thing I noticed on first look was that labels weirdly are range 1-10.. Oh 1-indexed languages, why do you exist :)

**Comment by uridah** [comment]
Working on it!

**Comment by soumith** [comment]
fixed via #98

**Comment by SoloyT** [comment]
Hi 
I want to load SHVN in python for training and testing. how can I use such this code for SVHN. ?
def load_mnist():
    # the data, shuffled and split between train and test sets
    from keras.datasets import mnist
    (x_train, y_train), (x_test, y_test) = mnist.load_data()
    x_train = x_train.reshape(-1, 28, 28, 1).astype('float32') / 255.
    x_test = x_test.reshape(-1, 28, 28, 1).astype('float32') / 255.
    y_train = to_categorical(y_train.astype('float32'))
    y_test = to_categorical(y_test.astype('float32'))
    return (x_train, y_train), (x_test, y_test)

Thanks!

**Comment by fmassa** [comment]
@SoloyT it looks like you want to use keras?

**Comment by SoloyT** [comment]
@fmassa Yes I prefer to use Keras, however I couldn't find keras code for that.

**Comment by fmassa** [comment]
I have no idea if keras support torch datasets, sorry.
Might be better asking in the Keras forums

**Comment by SoloyT** [comment]
ok, but it seems there is no solution via Keras. so would you please help me to load SVHN dataset without using keras?
Thanks

**Comment by fmassa** [comment]
I can't say it will work with keras, but here is how you'd do it for pytorch-like dataloaders:
```python
dataset = torchvision.datasets.SVHN('path/where/to/save', download=True)
```
to load the `train` split. More info can be [found in the documentation](https://pytorch.org/docs/stable/torchvision/datasets.html#svhn).
I won't be able to help you though if keras dataloader does not expect this format. Sorry

**Comment by SoloyT** [comment]
thank you so much @fmassa

**Comment by Ankita-Das** [comment]
Is using the format 2 of the dataset(SVHN) as efficient as using the format 1,
The images in format 2 contain cropped images,but a single image(which should actually contain just 1 digit) contains portions of other digits too.

Also,is deskewing necessary for the format 2 dataset?
